README.md
# Simple Rating

This plugin is rating widget that is using Icons.

## Getting Started
Add the module to your project ``pubspec.yaml``:
```yaml
...
dependencies:
 ...
 simple_rating: ^0.0.1
...
```
And install it using ``flutter packages get`` on your project folder. After that, just import the module and use it:

## Usage example

```dart
import 'package:flutter/material.dart';
import 'package:simple_rating/simple_rating.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double _rating = 0.0;
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Simple Rating example app'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              SimpleRating(
                useHalfRating: true,
                onChanged: (v) {
                  setState(() {
                    this._rating = v;
                  });
                },
                starCount: 5,
                rating: _rating,
                size: 48,
                color: Colors.deepPurple,
                borderColor: Colors.deepPurple,
                spacing: 0.0,
              ),
              Text(
                "${this._rating}",
                style: TextStyle(fontSize: 24, color: Colors.deepPurple),
              )
            ],
          ),
        ),
      ),
    );
  }
}

```
Enjoy!

## Constructor parameters
``` 
useHalfRating                   -   Whether to use half rating
onChanged                       -   Rating changed callback
starCount                       -   The maximum amount of stars
rating                          -   The current value of rating
size                            -   The size of a single star
color                           -   The body color of star
borderColor                     -   The border color of star
spacing                         -   Spacing between stars(default is 0.0)
icons                           -   Icon set of Empty, Full, Half
```

### Screenshots

![alt text](https://raw.githubusercontent.com/jaceshim/simple_rating/master/simple_rating.gif)

## Author
Developed by [Jace Shim (심천보)](https://www.facebook.com/jaceshim.kr)

## Contributing
Feel free to help!


CHANGELOG.md
## [0.0.1]
* initial release.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:simple_rating_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/simple_rating.dart
import 'package:flutter/material.dart';

typedef void RatingChangeCallback(double rating);

/// Simple Rating
class SimpleRating extends StatelessWidget {
  final RatingIcon icons;
  final int starCount;
  final double rating;
  final RatingChangeCallback onChanged;
  final Color color;
  final Color borderColor;
  final double size;
  final bool useHalfRating;
  final double spacing;

  SimpleRating(
      {this.icons = const RatingIcon(),
      this.starCount = 5,
      this.rating = 0.0,
      this.onChanged,
      this.color,
      this.borderColor,
      this.size,
      this.spacing = 0.0,
      this.useHalfRating = true}) {
    assert(this.rating != null);
  }

  Widget buildStar(BuildContext context, int index) {
    Icon icon;
    if (index >= rating) {
      icon = Icon(
        Icons.star_border,
        color: borderColor ?? Theme.of(context).primaryColor,
        size: size ?? 25.0,
      );
    } else if (rating.toInt() >= index + 1) {
      icon = Icon(
        Icons.star,
        color: color ?? Theme.of(context).primaryColor,
        size: size ?? 25.0,
      );
    } else {
      icon = Icon(
        Icons.star_half,
        color: color ?? Theme.of(context).primaryColor,
        size: size ?? 25.0,
      );
    }

    return GestureDetector(
      onTap: () {
        if (this.onChanged != null) onChanged(index + 1.0);
      },
      onHorizontalDragUpdate: (dragDetails) {
        RenderBox box = context.findRenderObject();
        var _position = box.globalToLocal(dragDetails.globalPosition);
        var i = _position.dx / size;
        var newRating = i.round().toDouble();
        if (useHalfRating) {
          newRating = _customRound(i);
        }
        if (newRating > starCount) {
          newRating = starCount.toDouble();
        }
        if (newRating < 0) {
          newRating = 0.0;
        }
        if (this.onChanged != null) onChanged(newRating);
      },
      child: icon,
    );
  }

  double _customRound(double val) {
    if (val <= 0) {
      return 0;
    }

    final int intVal = val.toInt();
    double tempVal = val - intVal;
    if (tempVal < 0.5) {
      return intVal.toDouble() + 0.5;
    } else {
      return (intVal + 1).toDouble();
    }
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      color: Colors.transparent,
      child: Wrap(
          alignment: WrapAlignment.start,
          spacing: spacing,
          children: List.generate(starCount, (index) => buildStar(context, index))),
    );
  }
}

/// Rating Icon
class RatingIcon {
  final IconData emptyIcon;
  final IconData halfIcon;
  final IconData fullIcon;

  const RatingIcon({this.emptyIcon, this.halfIcon, this.fullIcon});
}


android/src/main/java/jaceshim/simple_rating/SimpleRatingPlugin.java
package jaceshim.simple_rating;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** SimpleRatingPlugin */
public class SimpleRatingPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "simple_rating");
    channel.setMethodCallHandler(new SimpleRatingPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:simple_rating/simple_rating.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double _rating = 0.0;
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Simple Rating example app'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              SimpleRating(
                useHalfRating: true,
                onChanged: (v) {
                  setState(() {
                    this._rating = v;
                  });
                },
                starCount: 5,
                rating: _rating,
                size: 48,
                color: Colors.deepPurple,
                borderColor: Colors.deepPurple,
                spacing: 0.0,
              ),
              Text(
                "${this._rating}",
                style: TextStyle(fontSize: 24, color: Colors.deepPurple),
              )
            ],
          ),
        ),
      ),
    );
  }
}
